### PR TITLE
Code Review

### PR DESCRIPTION
**Pull request by pavliukvlad**

### CODE REVIEW

File: StudentList/Common/Extensions/ActivityExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,18 @@
+﻿using System.Threading.Tasks;
+using Android.App;
+using StudentList.Common.Dialogs;
+
+namespace StudentList.Extensions
+{
+    public static class ActivityExtensions
+    {
+        public static async Task RunMethodWithLoaderAsync(this Activity activity, Task task)
```
Comment (igor-leshkevych) [inline StudentList/Common/Extensions/ActivityExtensions.cs:9]: You handle Task, and how about Task<T>?

File: StudentList/Fragments/FilterStudentsFragment.cs [hunk 2]
```diff
@@ -53,7 +57,7 @@ public override void OnViewCreated(View view, Bundle savedInstanceState)
         {
             base.OnViewCreated(view, savedInstanceState);
 
-            if (this.studentFilter != default(StudentFilter))
+            if (this.studentFilter != StudentFilter.Default)
```
Comment (igor-leshkevych) [inline StudentList/Fragments/FilterStudentsFragment.cs]: You don't need a check here. If filter is a default filter, then you will simply have nulls in all fields, so nothing goes wrong if you set these nulls to textboxes.

File: StudentList/Services/PhotoService.cs [hunk 3]
```diff
@@ -4,44 +4,50 @@
...
             if (bitmap != null)
             {
                 try
                 {
-                    var root = context.GetDir("images", FileCreationMode.Private);
+                    var root = context.GetDir(StudentListDirNames.Image, FileCreationMode.Private);
                     var file = new Java.IO.File(root, fileName);
 
-                    using (var stream = new FileOutputStream(file))
+                    file.Delete();
```
Comment (igor-leshkevych) [inline StudentList/Services/PhotoService.cs]: Check if exists before delete.

File: StudentList/Fragments/StudentListFragment.cs [hunk 4]
```diff
@@ -1,39 +1,45 @@
...
 
         public override void OnCreate(Bundle savedInstanceState)
         {
             base.OnCreate(savedInstanceState);
 
-            this.repository = new StudentsRepository(this.Activity);
+            this.studentAdapter = new StudentAdapter();
+            this.loadingDialog = new LoadingDialog(this.Context);
+            this.repository = new StudentsRepository(
+                this.Activity,
```
Comment (igor-leshkevych) [inline StudentList/Fragments/StudentListFragment.cs]: You cant have a reference to activit inside your repo class (repo = pure .net class without references to android framework).

File: StudentList/Fragments/StudentListFragment.cs [hunk 5]
```diff
@@ -119,15 +118,18 @@ private void StudentAdapterItemClick(object sender, Student e)
...
+            await this.Activity.RunMethodWithLoaderAsync(this.PutStudentsToAdapter(StudentFilter.Default));
             this.studentsCountTextView.Text = string.Format(
-                CultureInfo.InvariantCulture, this.GetString(Resource.String.student_count_pattern), students.Count);
+               CultureInfo.InvariantCulture, this.GetString(Resource.String.student_count_pattern), this.studentAdapter.ItemCount);
+        }
 
-            this.loadingProgressBar.Visibility = ViewStates.Invisible;
+        private async Task PutStudentsToAdapter(StudentFilter studentFilter)
+        {
+            var students = await this.repository.GetStudentsAsync(studentFilter);
```
Comment (igor-leshkevych) [inline StudentList/Fragments/StudentListFragment.cs]: You wrapped wrong function with loader. You should've wrapped repo call, not call to PutStudentsToAdapter as PutStudentsToAdapter is a UI function, and repo call is async function.

File: StudentList/Fragments/StudentProfileFragment.cs [hunk 6]
```diff
@@ -83,7 +90,7 @@ public override View OnCreateView(LayoutInflater inflater, ViewGroup container,
         {
             if (!string.IsNullOrWhiteSpace(this.StudentId))
             {
-                this.selectedStudent = await this.studentRepository.GetStudentById(this.StudentId);
+                await this.Activity.RunMethodWithLoaderAsync(this.GetSelectedStudentAsync(this.StudentId));
```
Comment (igor-leshkevych) [inline StudentList/Fragments/StudentProfileFragment.cs]: Same here, only true async calls are wrapped in RunWithLoader(), not their wrapper-functions.

File: StudentList/Fragments/StudentProfileFragment.cs [hunk 7]
```diff
@@ -116,7 +121,7 @@ public override bool OnOptionsItemSelected(IMenuItem item)
             switch (item.ItemId)
             {
                 case Resource.Id.action_confirm:
-                    this.ConfirmAsync();
+                    this.Activity.RunMethodWithLoaderAsync(this.ConfirmAsync());
```
Comment (igor-leshkevych) [inline StudentList/Fragments/StudentProfileFragment.cs:127]: Same here

File: StudentList/Fragments/StudentProfileFragment.cs [hunk 8]
```diff
@@ -136,11 +141,13 @@ public override void OnActivityResult(int requestCode, int resultCode, Intent da
                 if (data.Data == null)
                 {
                     this.profilePhoto = (Bitmap)data.GetParcelableExtra("data");
+                    this.profilePhotoGuid = Guid.NewGuid();
```
Comment (igor-leshkevych) [inline StudentList/Fragments/StudentProfileFragment.cs]: this line and one below are same for both if-else branches, extract them outside if

File: StudentList/Fragments/StudentProfileFragment.cs [hunk 9]
```diff
@@ -200,9 +208,9 @@ private void OnBirthdateEditTextFocus(object sender, View.FocusChangeEventArgs e
             }
         }
 
-        private async void SaveButtonClickAsync(object sender, EventArgs e)
+        private async void ConfirmButtonClickAsync(object sender, EventArgs e)
         {
-            await this.ConfirmAsync();
+            await this.Activity.RunMethodWithLoaderAsync(this.ConfirmAsync());
```
Comment (igor-leshkevych) [inline StudentList/Fragments/StudentProfileFragment.cs:214]: Same with loader and async/await here.

File: StudentList/Fragments/StudentProfileFragment.cs [hunk 10]
```diff
@@ -216,31 +224,27 @@ private void DataSetPickerDialog(object sender, DatePickerDialog.DateSetEventArg
...
-        private async Task AddStudent(string name, string birthdate, string uni, string group, string phone)
+        private async Task AddStudent(SavingPhotoResult savingPhotoResult, string name, string birthdate, string uni, string group, string phone)
         {
-            var profilePhotoUri = await PhotoService.SavePhotoAsync(
-               this.profilePhoto,
-               string.Format(CultureInfo.InvariantCulture, "{0}{1}_profile_photo.png", name.ToLowerInvariant(), this.GetHashCode()),
-               this.Activity);
-
-            var validationResult = await this.studentRepository.AddNewStudentAsync(name, profilePhotoUri, birthdate, group, uni, phone);
+            var validationResult = await this.studentRepository.AddNewStudentAsync(name, savingPhotoResult, birthdate, group, uni, phone);
```
Comment (igor-leshkevych) [inline StudentList/Fragments/StudentProfileFragment.cs]: You don't need SavingPhotoResult here, just URI or null if failed.

File: StudentList/Fragments/StudentProfileFragment.cs [hunk 11]
```diff
@@ -252,15 +256,10 @@ private void DataSetPickerDialog(object sender, DatePickerDialog.DateSetEventArg
...
+        private async Task ChangeStudentById(string studentId, SavingPhotoResult savingPhotoResult, string name, string birthdate, string group, string uni, string phone)
         {
-            var profilePhotoUri = await PhotoService.SavePhotoAsync(
-                this.profilePhoto,
-                string.Format(CultureInfo.InvariantCulture, "{0}{1}_profile_photo.png", name.ToLowerInvariant(), this.GetHashCode()),
-                this.Activity);
-
             var validationResult = await this.studentRepository.ChangeStudentById(
-                this.StudentId, profilePhotoUri, name, birthdate, group, uni, phone);
+                this.StudentId, savingPhotoResult, name, birthdate, group, uni, phone);
```
Comment (igor-leshkevych) [inline StudentList/Fragments/StudentProfileFragment.cs]: Same here

File: StudentList/Fragments/StudentProfileFragment.cs [hunk 12]
```diff
@@ -278,15 +277,21 @@ private void SetErrors(ValidationResult validationResult)
...
                 foreach (var ctr in this.layouts)
                 {
-                    if (error.Key == ctr.Key)
+                    ctr.Value.Error = validationResult.GetErrorMessages(ctr.Key);
+
+                    if (error.Key == StudentProfileKeys.Photo)
                     {
-                        string errorMess = string.Join(".", error.Value);
-                        ctr.Value.Error = errorMess;
+                        Toast.MakeText(this.Context, string.Join(".", error.Value), ToastLength.Short).Show();
```
Comment (igor-leshkevych) [inline StudentList/Fragments/StudentProfileFragment.cs]: Showing Toast in foreach is a bad idea

File: StudentList/Fragments/StudentProfileFragment.cs [hunk 13]
```diff
@@ -278,15 +277,21 @@ private void SetErrors(ValidationResult validationResult)
...
                     }
                 }
             }
         }
 
+        private async Task GetSelectedStudentAsync(string studentId)
```
Comment (igor-leshkevych) [inline StudentList/Fragments/StudentProfileFragment.cs]: Useless function, you don't need a function to save data in one field

File: StudentList/Models/ValidationResult.cs [hunk 14]
```diff
@@ -13,5 +13,20 @@ public ValidationResult()
         public bool IsValid => !this.Errors.Any();
 
         public Dictionary<string, IEnumerable<string>> Errors { get; private set; }
+
+        public string GetErrorMessages(string key)
+        {
```
Comment (igor-leshkevych) [inline StudentList/Models/ValidationResult.cs:18]: use LINQ here, this code can be written in one line using it.

File: StudentList/Repositories/StudentsRepository.cs [hunk 15]
```diff
@@ -111,15 +110,20 @@ public StudentsRepository(Activity activity)
...
         }
 
-        private ValidationResult Validate(string name, string birthdate, string group, string uni, string phone)
+        private ValidationResult Validate(SavingPhotoResult photo, string name, string birthdate, string group, string uni, string phone)
         {
             var validationResult = new ValidationResult();
 
+            if (photo.IsError)
+            {
+                validationResult.Errors.Add(nameof(photo), new List<string> { this.activity.GetString(Resource.String.photo_toast_error) });
```
Comment (igor-leshkevych) [inline StudentList/Repositories/StudentsRepository.cs]: Don't use Context.GetString inside repository, it must be a pure .net class w/o refs on android.

**Review by igor-leshkevych — CHANGES_REQUESTED** [review]

File: StudentList/Common/Extensions/ActivityExtensions.cs (hunk 1 above)
Comment (pavliukvlad) [reply StudentList/Common/Extensions/ActivityExtensions.cs:9]: image
What do you think about that?

File: StudentList/Fragments/StudentProfileFragment.cs (hunk 7 above)
Comment (pavliukvlad) [reply StudentList/Fragments/StudentProfileFragment.cs:127]: But I have a several async/await methods in ConfirmAsync method and LoadingDialog should be shown when they are running. Currently, I have the one solving, I can just write smth that:
private async Task ConfirmAsync()
{ 
     loadingDialog.Show();
     //async/await method running
     loadingDialog.Hide()
}

So I don't know if this is correct. Do you have any ideas?

File: StudentList/Common/Extensions/ActivityExtensions.cs (hunk 1 above)
Comment (igor-leshkevych) [reply StudentList/Common/Extensions/ActivityExtensions.cs:9]: I think you will soon get tired of writing a new method for every new task result (List<Student>, Student, int, double, etc). Maybe you can simply use generic for that?)

File: StudentList/Fragments/StudentProfileFragment.cs (hunk 7 above)
Comment (igor-leshkevych) [reply StudentList/Fragments/StudentProfileFragment.cs:127]: The main idea is to wrap exactly a thing that you want to actually be covered by loader. Usually, it's some single long-running operation, that's why I recommended to wrap repo calls. But of course, if you have several long-running operations and you want them all to run with a same loader, then you can have a wrapper function (like ConfirmAsync) and surround it with loader.

File: StudentList/Repositories/StudentsRepository.cs [hunk 16]
```diff
@@ -115,40 +111,35 @@ public StudentsRepository(Activity activity, LoadingDelays loadingDelays)
...
 
-            if (photo.IsError)
-            {
-                validationResult.Errors.Add(nameof(photo), new List<string> { this.activity.GetString(Resource.String.photo_toast_error) });
-            }
-
             if (string.IsNullOrWhiteSpace(name))
             {
-                validationResult.Errors.Add(nameof(name), new List<string>() { this.activity.GetString(Resource.String.name_field_error) });
+                validationResult.Errors.Add(nameof(name), new List<string>() { "Name cannot be empty" });
```
Comment (igor-leshkevych) [inline StudentList/Repositories/StudentsRepository.cs]: And how you gonna have these strings localized?)

**Review by igor-leshkevych — CHANGES_REQUESTED** [review]
Hardcoded strings in repository. And moving that to constant's won't work, too, because you need to have an option of localizing this strings (same as Context.GetString()) does.

File: StudentList/Repositories/StudentsRepository.cs (hunk 16 above)
Comment (vladpavlyuk) [reply StudentList/Repositories/StudentsRepository.cs]: image
Could I write kind of that?

Comment (pavliukvlad) [reply StudentList/Repositories/StudentsRepository.cs]: I fixed that

File: .gitignore [hunk 17]
```diff
@@ -258,4 +258,5 @@ paket-files/
 
 # Python Tools for Visual Studio (PTVS)
 __pycache__/
-*.pyc
\ No newline at end of file
+*.pyc
+/StudentList.StudentList.apk
```
Comment (igor-leshkevych) [inline .gitignore:262]: Separate that with a comment from the rest of a file.

**Review by igor-leshkevych — APPROVED** [review]

**Pull request merged**